File: graph_generator.py
import random
import math


class Edge:
    def __init__(self, u, v, w):
        self.u = u  # Source
        self.v = v  # Destination
        self.w = w  # Weight


class Graph:
    def __init__(self, n, d):
        self.n = n  # Number of vertices
        self.d = d  # Graph dimension
        self.edges = []  # List of edges

    def create_graph(self):
        if self.d == 0:
            self.create_0D()
        elif self.d == 1:
            self.create_1D()
        else:
            self.create_234D()

    # Complete
    def create_0D(self):
        for i in range(1, self.n):
            for j in range(i):
                weight = random.random()  # CHECK CHECK CHECK – range is [0, 1)
                self.edges.append(Edge(i, j, weight))

    # Hypercube
    def create_1D(self):
        for i in range(1, self.n):
            for j in range(i):
                diff = i - j
                if diff & (diff - 1) == 0:  # Check if diff is a power of 2
                    weight = random.random()
                    self.edges.append(Edge(i, j, weight))

    # Geometric
    def create_234D(self):
        vertices = []  # Create list of n-dimensional vertices
        for i in range(self.n):
            vertex = []
            for j in range(self.d):
                vertex.append(random.random())
            vertices.append(vertex)

        for i in range(1, self.n):
            for j in range(i):
                weight = self.distance(vertices[i], vertices[j])
                self.edges.append(Edge(i, j, weight))

    # Calculate the Euclidean distance between two vertices (dimension = len(u))
    def distance(self, u, v):
        return math.sqrt(sum((u[i] - v[i]) ** 2 for i in range(len(u))))


def main():
    n = 10
    graphs = {
        "0D": Graph(n, 0),
        "Hypercube": Graph(n, 1),
        "2D": Graph(n, 2),
        "3D": Graph(n, 3),
        "4D": Graph(n, 4),
    }

    # Print sample edges
    for label, graph in graphs.items():
        graph.create_graph()
        print(
            f"Sample edges from {label} graph:",
            [(e.u, e.v, e.w) for e in graph.edges[:5]],
        )


if __name__ == "__main__":
    main()

# NOTE: include threshold function on edges? From "Hint" in the assignment

File: graph_generator.cpp
#include "graph_generator.h"
#include <iostream>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <random>

using namespace std;

// Edge Constructor
Edge::Edge(int src, int dest, double weight) : u(src), v(dest), w(weight) {}

// Graph Constructor
Graph::Graph(int numVertices, int dimension) : n(numVertices), d(dimension) {}

// Create Graph Function
void Graph::createGraph() {
    if (d == 0) {
        create_0D();
    } else if (d == 1) {
        create_1D();
    } else {
        create_234D();
    }
}

void Graph::create_0D() {
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<double> dist(0.0, 1.0);

    for (int i = 1; i < n; i++) {
        for (int j = 0; j < i; j++) {
            double weight = dist(gen);  // Generates a random weight
            edges.emplace_back(i, j, weight);
        }
    }
}

void Graph::create_1D() {
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<double> dist(0.0, 1.0);

    for (int i = 1; i < n; i++) {
        for (int j = 0; j < i; j++) {
            int diff = i - j;
            if ((diff & (diff - 1)) == 0) { // Check if diff is a power of 2
                double weight = dist(gen);
                edges.emplace_back(i, j, weight);
            }
        }
    }
}

void Graph::create_234D() {
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<double> dist(0.0, 1.0);

    vector<vector<double>> vertices(n, vector<double>(d));

    // Generate random vertices
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < d; j++) {
            vertices[i][j] = dist(gen);  // Assign random coordinate
        }
    }

    // Compute Euclidean distance
    for (int i = 1; i < n; i++) {
        for (int j = 0; j < i; j++) {
            double weight = distance(vertices[i], vertices[j]);
            edges.emplace_back(i, j, weight);
        }
    }
}

double Graph::distance(const vector<double>& u, const vector<double>& v) {
    double sum = 0.0;
    for (size_t i = 0; i < u.size(); i++) {
        sum += pow(u[i] - v[i], 2);
    }
    return sqrt(sum);
}

void printSampleEdges(Graph& graph, const string& label) {
    cout << "Sample edges from " << label << " graph:\n";
    for (size_t i = 0; i < min(graph.edges.size(), size_t(5)); i++) {
        cout << "(" << graph.edges[i].u << ", " << graph.edges[i].v 
             << ", " << graph.edges[i].w << ")\n";
    }
    cout << endl;
}

